### PR TITLE
add channel

### PR DESCRIPTION
**Pull request by BusyJay**
Channel is used to create call object. So after this pr, a working client can finally be implemented.

### CODE REVIEW

File: src/channel.rs [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+    /// Primary user agent: goes at the start of the user-agent metadata sent on each request.
+    pub fn primary_user_agent(mut self, agent: &str) -> ChannelBuilder {
+        let agent_string = format_user_agent_string(agent);
+        self.options
+            .insert(PRIMARY_USER_AGENT_STRING, Options::String(agent_string));
+        self
+    }
+
+    /// If enable, allow the use of SO_REUSEPORT if it's available (default true).
+    pub fn reuse_port(mut self, reuse: bool) -> ChannelBuilder {
```
Comment (siddontang) [inline src/channel.rs]: no reuse_address ?

File: src/channel.rs [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+}
+
+impl Drop for ChannelArgs {
+    fn drop(&mut self) {
+        unsafe { grpc_sys::grpcwrap_channel_args_destroy(self.args) }
+    }
+}
+
+struct ChannelInner {
+    _environ: Arc<Environment>,
```
Comment (siddontang) [inline src/channel.rs]: why "_" here?

File: src/channel.rs (hunk 1 above)
Comment (BusyJay) [inline src/channel.rs]: There is no configuration to enable it explicitly.

File: src/channel.rs (hunk 2 above)
Comment (BusyJay) [inline src/channel.rs]: To mute the unused warning.

File: src/channel.rs [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+const OPT_HTTP2_INITIAL_SEQUENCE_NUMBER: &'static [u8] = b"grpc.http2.initial_sequence_number\0";
+const OPT_SO_REUSE_PORT: &'static [u8] = b"grpc.so_reuseport\0";
+const OPT_SSL_TARGET_NAME_OVERRIDE: &'static [u8] = b"grpc.ssl_target_name_override\0";
+const PRIMARY_USER_AGENT_STRING: &'static [u8] = b"grpc.primary_user_agent\0";
+
+/// Ref: http://www.grpc.io/docs/guides/wire.html#user-agents
+fn format_user_agent_string(agent: &str) -> CString {
+    let version = env!("CARGO_PKG_VERSION");
+    let trimed_agent = agent.trim();
+    let val = if trimed_agent.trim().is_empty() {
```
Comment (disksing) [inline src/channel.rs]: Why trim trimed_agent?

File: src/channel.rs [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+}
+
+enum Options {
+    Integer(usize),
+    String(CString),
+}
+
+/// Channel configuration object.
+pub struct ChannelBuilder {
+    environ: Arc<Environment>,
```
Comment (disksing) [inline src/channel.rs]: env is a more common shorthand.

File: src/channel.rs [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+impl ChannelBuilder {
+    pub fn new(env: Arc<Environment>) -> ChannelBuilder {
+        ChannelBuilder {
+            env: env,
+            options: HashMap::new(),
+        }
+    }
+
+    /// Default authority to pass if none specified on call construction.
+    pub fn default_authority(mut self, authority: String) -> ChannelBuilder {
```
Comment (siddontang) [inline src/channel.rs]: `Into<String>` to support &str may be better here.

**Comment by disksing** [comment]
LGTM.

**Pull request merged**